File: ios/Mandalart/Vault/VaultWriteLedger.swift
import Foundation

/// 自分が最後に各 vault ファイルへ書いた content の SHA-256 を覚えておく台帳 (echo-skip / clobber 安全化、Stage ④)。
///
/// flush (DB→vault) は disk と DB の差分を見て DB 内容で上書きするため、アプリ前面中に外部編集が disk に届くと
/// 取り込み前の auto-flush がそれを潰す (iOS は watcher が無く取り込みは「背面=flush / 復帰=reconcile」の 2 点のみ)。
/// flush 直前に **disk の hash がこの台帳と違えば外部編集とみなして上書きを skip** することでこれを防ぐ。
///
/// reconcile (vault→DB 取り込み) は取り込み後に現 disk の hash を seed する。これが無いと Stage ③ の
/// 「取り込み → 次回 flush で frontmatter を本文に整合」書込み自体が外部編集と誤判定され永久 skip される。
///
/// process メモリのみ (永続化しない): cold start で必ず起動時 reconcile が現 disk から再 seed するため。
@MainActor
final class VaultWriteLedger {
    private var hashes: [String: String] = [:]   // key = dirName + "/" + path

    func storedHash(_ key: String) -> String? { hashes[key] }
    func record(_ key: String, hash: String) { hashes[key] = hash }
    func remove(_ key: String) { hashes.removeValue(forKey: key) }
}

/// 台帳キー。`VaultFile.path` は bare filename でマンダラート間で衝突するので dirName を畳む。
/// flush / reconcile 双方が同一に算出する (どちらも dirName と path を持つ)。
func vaultLedgerKey(dirName: String, path: String) -> String { dirName + "/" + path }

File: desktop/src/lib/vault/vaultWriteLedger.ts
/**
 * vault ファイルごとの「自分が最後に同期した content の hash」を持つ in-memory 台帳 (Phase 2 productize / clobber 安全化)。
 * iOS [VaultWriteLedger.swift](../../../../ios/Mandalart/Vault/VaultWriteLedger.swift) 相当。
 *
 * 1 つの台帳を 2 目的に使う:
 *  1. **flush の clobber ガード** (DB→vault): file を上書きする前に disk の現在 hash と照合し、
 *     不一致 (= 前回同期以降に外部編集された) なら **上書きしない** ([_vaultSync.flushDbToVault])。
 *  2. **watcher の echo-skip** (vault→DB): watcher 発火パスの disk hash が一致 (= 自分の書込みの反響)
 *     なら reconcile しない ([useVaultWatcher])。
 *
 * **更新タイミング**: flush 書込み時 (record) と reconcile 取り込み時 (record) の両方。これにより
 * 「外部編集が未取り込み」(skip して保護) と「同期済み」(安全に正準化書き戻し) を区別できる。
 *
 * key は **絶対パス** (`_mandalart.md` がフォルダ間で衝突するため相対 path は不可)。
 * プロセスメモリのみ (永続化しない)。起動時は空だが、bootstrap reconcile / 初回 export が record する。
 */

export type WriteLedger = {
  /** file を同期した (書いた or 取り込んだ) ことを記録する。 */
  record: (key: string, hash: string) => void
  /** 前回同期以降に外部で変更されたか。台帳に無い key は false (= 初回 export を妨げない)。 */
  isExternallyModified: (key: string, currentHash: string) => boolean
  /** 台帳に key が記録済みか。 */
  has: (key: string) => boolean
  /** 全消去 (テスト用)。 */
  clear: () => void
}

export function createWriteLedger(): WriteLedger {
  const hashByKey = new Map<string, string>()
  return {
    record(key, hash) {
      hashByKey.set(key, hash)
    },
    isExternallyModified(key, currentHash) {
      const known = hashByKey.get(key)
      return known !== undefined && known !== currentHash
    },
    has(key) {
      return hashByKey.has(key)
    },
    clear() {
      hashByKey.clear()
    },
  }
}

/** flush (_vaultSync) と watcher (useVaultWatcher) が共有する default singleton。 */
export const writeLedger = createWriteLedger()
